[wasm] Move Wasm app building targets to new WasmApp.targets

radical opened the pull request:
-    Move the uses of `WasmAppBuilder`, and
    `WasmLoadAssembliesAndReferences` tasks to a single targets file -
    `src/mono/wasm/build/WasmApp.targets`.

- Any project that wants to use that, can import the file, and set up the
   various properties before the target `WasmBuildApp` gets executed.

- `WasmBuildApp` is not run by default. The importing project will have
    to do that.

- By default, the `WasmLoadAssembliesAndReferences` task is not run, and
    the specified `@(WasmAssembliesToBundle)` are directly passed to
    `WasmAppBuilder`.
      - If the project needs assembly dependencies to be resolved, then
        set `$(WasmResolveAssembliesBeforeBuild) == true`.

- Assemblies to be bundled with the app are set via
    `@(WasmAssembliesToBundle)` (which optionally will have dependencies
    resolved)

The various task inputs correspond to properties as:

```
          AssemblySearchPaths               : @(WasmAssemblySearchPaths)
          Assemblies                        : @(WasmAssembliesToBundle)
          AppDir                            : $(WasmAppDir)

          MainAssembly                      : $(WasmMainAssemblyPath)
          MainJS                            : $(WasmMainJSPath)
          InvariantGlobalization            : $(WasmInvariantGlobalization)
          SatelliteAssemblies               : @(WasmSatelliteAssemblies)
          FilesToIncludeInFileSystem        : @(WasmFilesToIncludeInFileSystem)
          DebugLevel                        : $(WasmDebugLevel)
          MicrosoftNetCoreAppRuntimePackDir : $(MicrosoftNetCoreAppRuntimePackRidDir)
```

This should be a step towards eventually having this build as a sdk.

radical committed:
[wasm] Fix the name for MicrosoftNetCoreAppRuntimePackRidDir

We have `MicrosoftNetCoreAppRuntimePackDir=artifacts/bin/microsoft.netcore.app.runtime.browser-wasm/Release/`

.. and `MicrosoftNetCoreAppRuntimePack*Rid*Dir` as `$(MicrosoftNetCoreAppRuntimePackDir)/runtimes/browser-wasm/`.

But the wasm samples, and debugger test project files mixed up the
naming, though with correct paths. The following commit will move the
wasm build stuff to a single file, and this change will be useful to
have uniform, and correct naming.

radical committed:
[wasm] Don't quote multiple msbuild args

Else it ends up in the command line as:
    `dotnet build foo.proj "/p:TargetOS=Browser /p:Configuration=Release"`
instead of
    `dotnet build foo.proj /p:TargetOS=Browser /p:Configuration=Release`

radical committed:
[wasm] Add new `WasmApp.targets`

Move the uses of `WasmAppBuilder`, and
`WasmLoadAssembliesAndReferences` tasks to a single targets file -
`src/mono/wasm/build/WasmApp.targets`.

- Any project that wants to use that, can import the file, and set up the
various properties before the target `WasmBuildApp` gets executed.

- `WasmBuildApp` is not run by default. The importing project will have
to do that.

- By default, the `WasmLoadAssembliesAndReferences` task is not run, and
the specified `@(WasmAssembliesToBundle)` are directly passed to
`WasmAppBuilder`.
    - If the project needs assembly dependencies to be resolved, then
    set `$(WasmResolveAssembliesBeforeBuild) == true`.

- Assemblies to be bundled with the app are set via
`@(WasmAssembliesToBundle)` (which optionally will have dependencies
resolved)

The various task inputs correspond to properties as:

```
      AssemblySearchPaths               : @(WasmAssemblySearchPaths)

      Assemblies                        : @(WasmAssembliesToBundle)

      AppDir                            : $(WasmAppDir)
      MicrosoftNetCoreAppRuntimePackDir : $(MicrosoftNetCoreAppRuntimePackRidDir)
      MainAssembly                      : $(WasmMainAssemblyPath)
      MainJS                            : $(WasmMainJSPath)
      InvariantGlobalization            : $(WasmInvariantGlobalization)
      SatelliteAssemblies               : @(WasmSatelliteAssemblies)
      FilesToIncludeInFileSystem        : @(WasmFilesToIncludeInFileSystem)
      DebugLevel                        : $(WasmDebugLevel)
```

This should be a step towards eventually having this build as a sdk.

mdh1418 commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -0,0 +1,65 @@
+    <WriteLinesToFile
+      Condition="'$(WasmRunV8ScriptPath)' != ''"
+      File="$(WasmRunV8ScriptPath)"
+      Lines="v8 --expose_wasm runtime.js -- --run $(MainAssembly) $*"
```
I think $(MainAssembly) needs to be resolved to the filename of `$(WasmMainAssemblyPath)` with `.dll` extension if `$(WasmRunV8ScripPath)` is set.

radical replied:
Good catch, fixed.

mdh1418 commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -0,0 +1,65 @@
+  <UsingTask TaskName="WasmLoadAssembliesAndReferences" AssemblyFile="$(WasmAppBuilderTasksAssemblyPath)" />
+
+  <Target Name="WasmBuildApp" AfterTargets="Publish">
+    <!-- FIXME: TArgetOS check -->
```
Can the check be added in this PR?

radical replied:
Oops, forgot to remove this. I'm leaving the check up to the project importing this file.

radical replied:
I added an `Error` to catch a case where this file was included incorrectly. Thanks.

radical committed:
[wasm] Use the correct assembly path property

radical committed:
Check for TargetOS==Browser

radical committed:
Pass remaining WasmAppBuilder args - IcuDataFileName, and RemoteSources

radical committed:
WasmAppBuilder: don't emit run-v8.sh in the task

radical committed:
WasmTestRunner: use the new WasmApp.targets

radical committed:
Include WasmMainAssemblyPath for dependency resolution, by default

radical committed:
Remove unused run-v8.sh generation

radical commented:
I have removed instances of `run-v8.sh` being generated, where it isn't needed, AFAICS!

```
$ git grep run-v8.sh
src/tests/Common/CLRTest.Execute.Bash.targets:./run-v8.sh
src/tests/run.proj:      <AllCMDsPresent Remove="$(_CMDDIR)\**\run-v8.sh" Condition="'$(TargetArchitecture)' == 'wasm'" />
src/tests/run.proj:      <AllRunnableTestPaths Remove="$(XunitTestBinBase)\**\run-v8.sh" Condition="'$(TargetArchitecture)' == 'wasm'" />
```

I'm not sure how I can test the above `CLRTest.Execute.Bash.targets` case though. Hopefully, I didn't break anyone.

vargaz commented:
run-v8.sh is a helper script to be used during development it shouldn't removed imho.

radical committed:
Add back run-v8.sh for console sample, WasmTestRunner, and lib tests

radical committed:
WasmApp.targets: Use a default path for run-v8.sh

.. and control it's generation through `$(WasmGenerateRunV8Script)`

radical committed:
run-v8.sh: use only the filename on the command line

radical committed:
update README.md

lewing commented on eng/testing/tests.mobile.targets:
```diff
@@ -195,5 +192,5 @@
           Condition="'$(IsCrossTargetingBuild)' != 'true'"
           AfterTargets="Build"
-          DependsOnTargets="Publish;BundleTestAppleApp;BundleTestAndroidApp;BundleTestWasmApp;ArchiveTests" />
+          DependsOnTargets="Publish;BundleTestAppleApp;BundleTestAndroidApp;WasmBuildApp;ArchiveTests" />
```
It looks like you are only including the targets file on Browser to the WasmBuildApp target doesn't exist on other platforms and things break.

radical replied:
Yikes! Hopefully, fixed with the new commit.

lewing reviewed (commented):
Looks good but needs a fix for the other platforms

radical committed:
[wasm] Make the wasm targets conditional on `TargetOS==Browser`

.. instead of conditionally importing `WasmApp.targets`. This allows
other targets to depend on wasm targets, and execute only if it makes
sense.

Based on @lewing's feedback.

radical committed:
[wasm][tests] fix check for when to generate run-v8.sh

radical committed:
WasmTestRunner.proj: fix typo

radical committed:
[wasm] Always generate the run-v8.sh script, for browser runs

The `Scenario==WasmTestOnBrowser` seems to fail for JIT tests, which
don't use the browser.

radical committed:
[wasm] Fix tests

The real issue seems to be that the helix bundle doesn't have
`WasmApp.targets`. Fixing that, which should hopefully fix the tests!

radical committed:
[wasm] Another attempt at fixing the tests

Mimic earlier behavior of having a `BundleTestWasmApp`, and making it
conditional on `TargetOS == Browser`.

radical committed:
fix property reference

radical committed:
[wasm] skip missing assemblies for the jit tests, same as the original code

mdh1418 commented on eng/testing/tests.mobile.targets:
```diff
@@ -146,38 +146,37 @@
 
-  <UsingTask TaskName="WasmAppBuilder"
-             AssemblyFile="$(WasmAppBuilderTasksAssemblyPath)" />
+  <Import Project="$(MonoProjectRoot)\wasm\build\WasmApp.targets" Condition="'$(TargetOS)' == 'Browser'" />
```
Could this be moved to the bottom (convention of putting target imports at the end of the file)?

radical replied:
I can get this in a follow up PR.

mdh1418 commented on eng/testing/tests.mobile.targets:
```diff
@@ -146,38 +146,37 @@
+      <WasmMainAssemblyPath>$(PublishDir)WasmTestRunner.dll</WasmMainAssemblyPath>
+      <WasmMainJSPath>$(MonoProjectRoot)\wasm\runtime-test.js</WasmMainJSPath>
+      <WasmInvariantGlobalization>$(InvariantGlobalization)</WasmInvariantGlobalization>
+      <WasmGenerateRunV8Script>true</WasmGenerateRunV8Script>
```
Is it certain that `'$(Scenario)' != 'WasmTestOnBrowser'` caused those Runtime tests failures? I saw them on https://github.com/dotnet/runtime/pull/45907 as well but wasn't certain because I was under the assumption that `run-v8.sh` was only used locally (not on CI) to help with development.

radical replied:
Tests from `src/tests` seem to use it too, via `src/tests/Common/CLRTest.Execute.Bash.targets`, and `WasmTestRunner`.

mdh1418 reviewed (approved):
Thanks for the consolidation of all the task calls! I find this reorganization very helpful :)
The changes look good to me.

radical committed:
Merge remote-tracking branch 'origin/master' into wasm-app

radical commented:
The single test failure is https://github.com/dotnet/runtime/issues/44657 .
